hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java

```diff
@@ -0,0 +1,308 @@
+			Serializable id,
+			SharedSessionContractImplementor session) throws HibernateException {
+		if ( isInverse ) {
+			return CompletionStages.voidFuture();
+		}
+
+		if ( !isRowDeleteEnabled() ) {
+			return CompletionStages.voidFuture();
+		}
+
+		CompletionStage<Integer> loop = CompletionStages.zeroFuture();
+		ReactiveConnection reactiveConnection = getReactiveConnection( session );
+
+		boolean deleteByIndex = hasIndex && !indexContainsFormula;
+
+		return CompletionStages.total(
+				entryArray( collection, deleteByIndex ),
+				entry -> loop.thenCompose( total -> reactiveConnection
+						.update( getSQLDeleteRowString(), paramValues( entry, id ,session, deleteByIndex ) )
+						.thenApply( deleteCount -> total + deleteCount ) ) )
+				.thenAccept( total -> LOG.debugf( "Done removing rows: %s removed", total ) );
+	}
+
+	/**
+	 * @see org.hibernate.persister.collection.AbstractCollectionPersister#updateRows(PersistentCollection, Serializable, SharedSessionContractImplementor)
+	 */
+	@Override
+	public CompletionStage<Void> reactiveUpdateRows(
+			PersistentCollection collection,
+			Serializable id,
+			SharedSessionContractImplementor session) throws HibernateException {
+
+		if ( !isInverse && collection.isRowUpdatePossible() ) {
+			// TODO:  convert AbstractCollectionPersister.doUpdateRows() to reactive logic
+			// update all the modified entries
+			// NOTE this method call uses a JDBC connection and will fail for Map ElementCollection type
+			// Generally bags and sets are the only collections that cannot be mapped to a single row in the database
+			// So Maps, for instance will return isRowUpdatePossible() == TRUE
+
+			throw new UnsupportedOperationException();
+		}
+		return CompletionStages.voidFuture();
+	}
+
+	/**
+	 * @see org.hibernate.persister.collection.AbstractCollectionPersister#insertRows(PersistentCollection, Serializable, SharedSessionContractImplementor)
+	 */
+	@Override
+	public CompletionStage<Void> reactiveInsertRows(
+			PersistentCollection collection,
+			Serializable id,
+			SharedSessionContractImplementor session) throws HibernateException {
+		if ( isInverse ) {
+			return CompletionStages.voidFuture();
+		}
+
+		if ( !isRowDeleteEnabled() ) {
+			return CompletionStages.voidFuture();
+		}
+
+		if ( LOG.isDebugEnabled() ) {
+			LOG.debugf(
+					"Inserting rows of collection: %s",
+					MessageHelper.collectionInfoString( this, collection, id, session )
+			);
+		}
+
+		collection.preInsert( this );
+
+		CompletionStage<Integer> loop = CompletionStages.zeroFuture();
+		ReactiveConnection reactiveConnection = getReactiveConnection( session );
+
+		return CompletionStages.total(
+				entryArray(collection),
+				entry -> loop.thenCompose( total -> reactiveConnection
+						.update( getSQLInsertRowString(), paramValues( entry, collection, id ,session ) )
+						.thenApply( insertCount -> total + insertCount ) ) )
+				.thenAccept( total -> LOG.debugf( "Done inserting rows: %s inserted", total ) );
```

Comment:
I haven't tested it, but I think this should be:
```
		return CompletionStages
				.total( entryArray( collection ), 
						entry -> reactiveConnection.update( getSQLInsertRowString(), paramValues( entry, collection, id,session ) ) )
				.thenAccept( total -> LOG.debugf( "Done inserting rows: %s inserted", total ) );
```

Reply:
By the way, if you create a `total` method in `CompletionStages` that accepts an Iterator or Iterable you can remove the conversion you need to do in `entryArray`